### PR TITLE
Simplify when-equations

### PR DESCRIPTION
**Pull request by henrikt-ma**

### CODE REVIEW

**Review by HansOlsson — CHANGES_REQUESTED** [review]
I believe it is generally good, but will need to consider it more - but with one caveat.
However, at first it may be good to clarify _why_ Modelica has this special rule - and the reason is that 

 ```
 when {initial(), some event} then
     some complicated equations
  end when;
```
Have two consequences:
* It is the only way to ensure that those same equations are activated both initially and at some event.
* It replaces the pre(x)=x equation that is otherwise introduced during initialization (to avoid over-determined initialization); https://specification.modelica.org/master/equations.html#initialization-initial-equation-and-initial-algorithm

The latter means that there is an ambiguity for handling this as
```
   when {initial()} then
      x=2;
   end when
```;
is mapped to
```
initial equation
  x=2;
```
and then either nothing (which gives no equation during initialization) or:
```
   when fill(false,0) then
      x=2;
   end when
```
which during initialization gives `x=pre(x)`. There are a number of ways of handling this, but currently it is unclear.

**Comment by henrikt-ma** [comment]
Good points.  What I had in mind was a semantics for `when`-equations where they are simply removed completely from the initialization problem, meaning that the `x = pre(x)` should be added when appropriate as full Modelica is reduced to Flat Modelica.

**Comment by olivleno** [comment]
Web Meeting: Gerd, Hans, Henrik, Martin, Oliver

Hans: We need to consider not only `when initial()` but also cases `when` without `initial()`, which yields `x=pre(x)` in the initial equations.

Henrik: We can force the default equations to be explicit in flat Modelica by saying that when equations have no meaning at all for the initialization problem. 

Henrik: when statements triggered by initial() can be converted into if statements. Hence, in flat Modelica they should be required to be converted.

Hans: In full Modelica there are semantic restrictions on when statements and we need to ensure that they are appreciated also after conversion to if statements.
Especially consequences from moving when statement content to if statements related to changed discrete time variability. To be continued.

 * Prepare a concrete example for this case [Henrik]

**Comment by HansOlsson** [comment]
Apart from pre (and related operators like edge and change) I thought that the main consequence of having when-code outside of when-statement would be that we get new event-generation.

At first I thought it didn't matter, but thinking more I see a larger issue - and my conclusion is that either we don't change this, or we have some special construct (not sure if "if initial()" suffices).

I haven't looked at all the consequences yet of events during initialization, but I started with the following:

```
algorithm
  when {b, initial()} then
    b2 = x > 0;
  end when;
```

My first idea to avoid the event-generation was:
```
algorithm
  when {b} then
    b2 = x > 0;
  end when;
   if initial() then
      b2 = noEvent(x>0);
   end if;
```
But you cannot assign a non-discrete time boolean value (such as `noEvent(x>0)`) to a normal boolean variable.

That indicates that I don't know how to handle:
```
algorithm
  when {b, initial()} then
    b2 = noEvent(x>0);
  end when;
```

File: RationaleMCP/0031/differences.md [hunk 1]
```diff
@@ -151,6 +151,84 @@ end 'M';
...
+Note that the following is illegal since the argument of `pre` needs to be discrete-time:
+```
+  if initial() then
+    'y' := pre('x'); /* 'x' is continuous-time, since no longer inside when-clause. */
+  end if;
+```
+Open questions:
+- What are the other ways one might depend on the discrete-time variability inside a `when`-clause?
+- Could there be a resolution in utilizing that `pre('x')` appears guarded by `initial()` in the `if`-statement, as this means that `pre('x')` refers to an independent variable of the initialization problem?
+- Better to simply allow `when`-statements to be activated during initialization by `initial()` in Flat Modelica?
```
Comment (henrikt-ma) [inline RationaleMCP/0031/differences.md:229]: Commenting in responds to this comment by @HansOlsson:
> At first I thought it didn't matter, but thinking more I see a larger issue - and my conclusion is that either we don't change this, or we have some special construct (not sure if "if initial()" suffices).

Yes, this is very similar to what I concluded with the open questions left here.  Right now, if we don't have ideas to try, I'd say there's not enough time to try to do anything about the `when`-statements.  I think we should just leave a comment that unlike `when`-equations, there is no difference between full Modelica and Flat Modelica when it comes to `when`-statements.

Comment (olivleno) [reply RationaleMCP/0031/differences.md:229]: Web Meeting (Hans, Henrik, Gerd, Martin, Oliver)

Discussed the when-equation and when-statement examples:
* For when-equations the `when initial()` case can be simplified as proposed.
* For when-statements the discussion confirmed that there is no straight-forward solution.

Conclusion:
Leave the text for when-equations as is.
Replace the text for when-statements with a comment that no differences w.r.t. to full Modelica are applied in flat Modelica. A comment shall explain that the case above is not as easily solved as for when-equations.

Decision:
Fix the PR accordingly [Henrik]
Approve in next meeting.

**Pull request merged**